File: 2019/src/main/scala/day01/day1.scala
package day01

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source

def parseInput(input: List[String]) = input.map(_.toInt)

def helper(weights: List[Int]) = weights.map(_ / 3 - 2).filter(_ > 0)

def evaluatorOne(weights: List[Int]): Int = helper(weights).sum

def evaluatorTwo(weights: List[Int]): Int = {
    return Iterator.iterate(weights)(helper).takeWhile(_.nonEmpty).drop(1).flatten.sum
}

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day01.txt") match {
        case Success(lines) => {
            val input = parseInput(lines)
            println(s"Part One: ${evaluatorOne(input)}")
            println(s"Part Two: ${evaluatorTwo(input)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}

File: 2019/src/main/kotlin/day01/day1.kt
package day01

import java.io.File

fun parseInput(lines: List<String>) = lines.map { it.toInt() }

fun helper(weights: List<Int>) = weights.map { it / 3 - 2 }.filter { it > 0 }

fun evaluatorOne(weights: List<Int>) = helper(weights).sum()

fun evaluatorTwo(weights: List<Int>): Int {
    var fuels = helper(weights)
    var total = 0

    while (fuels.isNotEmpty()) {
        total += fuels.sum()
        fuels = helper(fuels)
    }

    return total
}

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/day01.txt")
        .onSuccess {
            val weights = parseInput(it)
            println("Part One: ${evaluatorOne(weights)}")
            println("Part Two: ${evaluatorTwo(weights)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}